File: dailyquest_web/src/main/java/dailyquest/user/service/UserService.java
package dailyquest.user.service;

import dailyquest.common.DateTimeExtensionKt;
import dailyquest.common.MessageUtil;
import dailyquest.quest.entity.QuestType;
import dailyquest.redis.service.RedisService;
import dailyquest.user.dto.UserExpAndGoldRequest;
import dailyquest.user.dto.UserPrincipal;
import dailyquest.user.dto.UserSaveRequest;
import dailyquest.user.dto.UserUpdateRequest;
import dailyquest.user.entity.ProviderType;
import dailyquest.user.entity.UserInfo;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

@Transactional
@RequiredArgsConstructor
@Service
public class UserService {
    private final UserQueryService userQueryService;
    private final UserCommandService userCommandService;
    private final RedisService redisService;

    public UserInfo getOrSaveUser(String oauth2Id, ProviderType providerType) {
        UserInfo foundUser = userQueryService.findUser(oauth2Id);
        boolean needSave = foundUser == null;
        if (needSave) {
            String randomNickname = this.createRandomNickname();
            UserSaveRequest saveRequest = new UserSaveRequest(oauth2Id, randomNickname, providerType);
            foundUser = userCommandService.saveUser(saveRequest);
        }
        return foundUser;
    }

    private String createRandomNickname() {
        String candidateNickname = redisService.createRandomNickname();
        while (userQueryService.isDuplicateNickname(candidateNickname)) {
            candidateNickname = redisService.createRandomNickname();
        }
        return candidateNickname;
    }

    public UserInfo getUser(Long userId) {
        return userQueryService.findUser(userId);
    }

    public UserPrincipal getUserPrincipal(Long userId) {
        return UserPrincipal.create(this.getUser(userId), redisService.getExpTable());
    }

    public void updateUser(UserPrincipal principal, UserUpdateRequest updateRequest) throws IllegalStateException {
        UserInfo updateTarget = userQueryService.findUser(principal.getId());
        boolean updateFailed = !userCommandService.updateUser(updateTarget, updateRequest);
        if (updateFailed) {
            LocalDateTime updateAvailableTime = updateTarget.getUpdateAvailableDateTimeOfCoreTime();
            String timeSinceNowUntilAvailable = DateTimeExtensionKt.timeSinceNowAsString(updateAvailableTime);
            String errorMessage = MessageUtil.getMessage("user.settings.updateLimit", timeSinceNowUntilAvailable);
            throw new IllegalStateException(errorMessage);
        }
    }

    public void giveExpAndGoldToUser(QuestType type, UserInfo user) {
        UserInfo foundUser = userQueryService.findUser(user.getId());
        long questClearExp = redisService.getQuestClearExp();
        long questClearGold = redisService.getQuestClearGold();
        UserExpAndGoldRequest updateRequest = UserExpAndGoldRequest.of(type, questClearExp, questClearGold);
        userCommandService.updateUserExpAndGold(foundUser, updateRequest);
    }
}

File: user/src/main/kotlin/dailyquest/user/service/UserService.kt
package dailyquest.user.service

import dailyquest.achievement.entity.AchievementType.*
import dailyquest.common.timeSinceNowAsString
import dailyquest.user.dto.UserResponse
import dailyquest.user.dto.UserSaveRequest
import dailyquest.user.dto.UserUpdateRequest
import dailyquest.user.entity.User
import dailyquest.user.record.service.UserRecordService
import dailyquest.user.repository.UserRepository
import jakarta.persistence.EntityNotFoundException
import org.springframework.context.support.MessageSourceAccessor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import kotlin.jvm.optionals.getOrNull

@Transactional(readOnly = true)
@Service
class UserService(
    private val userRepository: UserRepository,
    private val userRecordService: UserRecordService,
    private val messageSourceAccessor: MessageSourceAccessor
) {
    fun findUserByOauthId(oauth2Id: String): UserResponse? {
        return userRepository.findByOauth2Id(oauth2Id)?.let { UserResponse.from(it) }
    }

    fun getUserById(userId: Long): UserResponse {
        return UserResponse.from(this.findUser(userId))
    }

    @Throws(EntityNotFoundException::class)
    private fun findUser(userId: Long): User {
        val foundUser = userRepository.findById(userId).getOrNull()
        val entityNotFoundMessage = messageSourceAccessor.getMessage("exception.entity.notfound", arrayOf(messageSourceAccessor.getMessage("user")))
        return foundUser ?: throw EntityNotFoundException(entityNotFoundMessage)
    }

    fun isDuplicatedNickname(nickname: String): Boolean {
        return userRepository.existsByNickname(nickname)
    }

    @Transactional
    fun saveUser(saveRequest: UserSaveRequest): Long  {
        val requestEntity = saveRequest.mapToEntity()
        val userId = userRepository.save(requestEntity).id
        userRecordService.saveNewRecordEntity(userId)
        return userId
    }

    @Transactional
    @Throws(IllegalStateException::class)
    fun updateUser(userId: Long, updateRequest: UserUpdateRequest) {
        val updateTarget = this.findUser(userId)
        val updateSucceed = updateTarget.updateUser(updateRequest)
        if (!updateSucceed) {
            val updateAvailableTime: LocalDateTime = updateTarget.getUpdateAvailableDateTimeOfCoreTime()
            val timeSinceNowUntilAvailable: String = updateAvailableTime.timeSinceNowAsString()
            val errorMessage: String = messageSourceAccessor.getMessage("user.coreTime.updateLimit", arrayOf(timeSinceNowUntilAvailable))
            throw IllegalStateException(errorMessage)
        }
    }

    @Transactional
    fun addUserExpAndGold(userId: Long, updateRequest: UserUpdateRequest) {
        val updateTarget = this.findUser(userId)
        updateTarget.addExpAndGold(updateRequest.earnedExp, updateRequest.earnedGold)
    }
}
